### ISSUE TITLE
"Save Project Data as Prawn Data file..." does not yield SQ2.50-readable Prawn XML

### ISSUE THREAD

**Issue by sbodorkos** [CIRDLES/Squid#574]
I have an interest in processing the Prawn XML file used in the Squid3 Demo Project, in SQUID 2.50. (Essentially because I want to verify that Squid3's arithmetic for varied Nscans inside a single Prawn XML file is working correctly.)

So I opened the Demo Project, visited the Data menu, selected "Save Project Data as Prawn Data file..." and saved the XML file. But when I attempt to parse (or even "condense") that file in SQUID 2.50, it fails. And I am assuming that in theory, it should not fail (if that's wrong, please let me know!).

It's very likely that this is SQUID 2.50's fault: I suspect that Ludwig's parser essentially treats the XML as a text-file (which is not the way you're supposed to do it, I know). Maybe the version Saved by Squid3 is different to the primary file generated by the SHRIMP control software in some way that is functionally insignificant in XML terms, but which would break text-style parsing? 

I'll try to do some tests (comparing initial Prawn XML with that exported by Squid3) using a Prawn XML I already have, but if you have a better idea, please add it here...

**Comment by cwmagee** [comment]
Ken's 2.5  can't deal with unknown tags. Poxi has the same problem.

**Comment by sbodorkos** [comment]
Yes, but I was interested in the case where I take a Prawn XML file generated by the SHRIMP control, software (which SQUID 2.50 obviously reads), bring it into a Squid3 Project, then do "Save Project Data as Prawn Data file...". Conceptually, I would expect Squid3 to give back a functionally unmodified XML file which SQUID 2.50 can still read.

Right now, it does not. However, I have now stepped through the VBA code that deals with Prawn XML parsing, hoping that I could find and fix the parsing errors sequentially (and that there would be a small number of them!). And once I got the hang of diagnosing the VBA crash-outs, it worked pretty well.

Here are the three Prawn XML files: 

- 01 is the original produced by the SHRIMP control software.
- 02 is the file written by Squid3, after adding 01 to a Squid3 Project and then selecting "Save Project Data as Prawn Data file..." from the Data menu, and which **cannot** be parsed by SQUID 2.50.
- 03 is a version of 02, which I edited in Notepad++ to address its incompatibilities with Ludwig's VBA code, and which **can** be parsed by SQUID 2.50.

**[836_1_PrawnXMLs_Original+Squid3+Sq3RepairedNotepad.zip](https://github.com/CIRDLES/Squid/files/5892082/836_1_PrawnXMLs_Original%2BSquid3%2BSq3RepairedNotepad.zip)**

The good news is that I only had to fix 3 types of "error" (in the sense of SQUID 2.50 parsing) in the Squid3-saved Prawn XML, before it became SQUID 2.50-readable. And I was able to address all three in Notepad++, so they're not complicated. My hope is that we can address these on the Squid3 side, because there is no "cost" on that side. I am not keen on editing the SQUID 2.50 code to accept the Squid3-generated XMLs: at best it would be messy, at worst I'd be very nervous about unintended consequences.

Essentially the SQUID 2.50-related issue is the one I flagged in the original post: the XML is being read as text (and the tags are being trimmed using Ludwig's text-editing functions, which are dangerous, and which I don't want to mess with). The issues and my solutions:

1. **Indents:** Squid3-saved Prawn XML files feature lots of pretty indenting when opened in Notepad++, handled by sets of quadruple-spaces (see image below). But these are absent from the original Prawn XML files (and thus not handled by Ludwig's text-searchers and text-trimmers), plus these space-sets also add about 25% to the file size. I performed a Find-and-Replace in Notepad++, where Find = 4 consecutive white spaces and Replace = <null>. This did the trick with respect to eliminating this start-of-line of error in the file-parsing, but it is possible that there were other (non-indent related) sets of quadruple-spaces elsewhere in the Prawn XML file which I should have retained but did not. I hope the Squid3-based solution can simply remove the indents.

![image](https://user-images.githubusercontent.com/17473203/106247772-1eec8200-6264-11eb-8e04-c70f64d09cbe.png)

2. **End-of line formatting for rows containing values:** Unfortunately, Ludwig has explicitly coded a white space in his end-of-line character-array for rows containing values, and the Squid3-saved Prawn XML files lack that white space. It is best described by example: the Squid3-saved file contains 76,437 occurrences of the 3-character sequence "slash greater-than linefeed", each of which is preceded by a double-quote (verified by Finding the 4-character sequence "double-quote slash greater-than linefeed", which also has 76,437 occurrences). The fix required for SQUID 2.50 parsing is to replace this 4-character sequence with the 5-character sequence "double-quote WHITE-SPACE slash greater-than linefeed". See image below. Hopefully it's trivial to add to the Squid3 side.

![image](https://user-images.githubusercontent.com/17473203/106246830-d84a5800-6262-11eb-821d-238e16c78683.png)

3. **Start-of-file Comment:** It turns out that SQUID 2.50 requires the Prawn XML to contain a line like Line 2 of the original file (see image below). Ludwig's parser expressly seeks the leading pair of characters ("less-than exclamation-mark") as a trigger to read and store the line, trim off the flanking hyphens etc., store the text as "XML file type", and write the result in cell C1 of the SQUID 2.50 condensed datasheet. Obviously none of this matters on the Squid3 side, and Squid3-saved XMLs do not currently contain this line. Unfortunately, if it is not present in the first few lines of the Prawn XML file, Ludwig's row-counter loses itself, and file-parsing fails. Obviously this is sub-optimal on the SQUID 2.50 side, but I am reluctant to edit that code. Plus the fix on the Squid3 side is simple: write Line 2 of the original file as a **static** Line 2 in Squid3-saved Prawn XML files. The static text could even be (universally) modified to read "" or similar, which might be a useful little piece of provenance information for anyone taking these Squid3-saved Prawn XML files back to SQUID 2.50.

![image](https://user-images.githubusercontent.com/17473203/106247250-5f97cb80-6263-11eb-85fa-ac42dcf70cd0.png)

**Comment by cwmagee** [comment]
The whole point of less-than exclamation point is that it indicates a comment which xml parsers should ignore. If Ken deliberately makes his code do the opposite of what it is supposed to, God help us all.

**Comment by sbodorkos** [comment]
Well, quite... I have wondered whether Ken did it deliberately, to make sure he was parsing a genuine SHRIMP SW-generated Prawn XML, rather than imitations of unknown origin (like we're trying to build in Squid3!). That last issue is easy enough to step around in the SQUID 2.50 code, in a "do no harm" way, and I will add the line of code to do that.

The other two are more scary, because any SQUID 2.50 code change that accommodates Squid3-generated XML must obviously not compromise current functionality in reading SHRIMP SW-generated ones. Not easy, unless I add code to SQUID 2.50 that explicitly seeks to identify Squid3-generated XML files and treat them differently. I'm not keen to do that - seems like a retrograde step, even if I get it right. 

I am hoping it is easier to accommodate both of those changes to XML on the Squid3 side, assuming that Squid3 users are not all that concerned about exactly what their XML files look like in a text-editor... Jim @bowring , your thoughts?

**Comment by cwmagee** [comment]
Does <! Bite me Ken !> work?

**Comment by bowring** [comment]
Squid3 in v 1.7.1 will output prawn xml files with the three changes requested by @sbodorkos and will then be compatible with both Squid25 and Squid3 for reading, pending confirmation tests by @sbodorkos, as I don't have Squid25.

**Comment by bowring** [comment]
@sbodorkos - attached is zip of revised output - test it with Squid25.  Note I added a second comment per your provenance suggestion that may break it for Squid25.  If not, we could make it even more informative.:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<!-- SHRIMP SW PRAWN Data File -->
<!-- SQUID3-generated PRAWN Data File -->
...
```

Thanks!
[ILC I_1_2012_FEB_08_10.38-REV.xml.zip](https://github.com/CIRDLES/Squid/files/5898147/ILC.I_1_2012_FEB_08_10.38-REV.xml.zip)

**Comment by sbodorkos** [comment]
Thanks Jim @bowring , when I unzipped and ran that XML unmodified (i.e. with both Comments in place, lines 2 and 3) through SQUID 2.50's "condenser", this is what I got:

![image](https://user-images.githubusercontent.com/17473203/106369952-e22aa300-63a9-11eb-82e9-5f4bcb8205fa.png)

Looks perfect. It also looks like Ken's text-trimmers are not unduly dependent on the length of the string being read for storage and display at C1, so feel free to add extra info if you wish.

But otherwise, I think our work here is done!
